### PR TITLE
Add two reduction schemas for sin terms

### PR DESCRIPTION
**Pull request by ajreynol**
This restores our reasoning for symmetry of sine, now on-demand and without introducing sin terms eagerly.

It also ensures that sin terms whose arguments fall exactly on boundary points are removed from consideration from the transcendental solver.  This allows us to answer sat for inputs involving sin(k) when k = pi/2.

It also cleans the relationship of sine solver and transcendental state, and makes a small fix to the monotonicity of sin schema.

### CODE REVIEW

File: src/theory/arith/nl/transcendental/sine_solver.cpp [hunk 1]
```diff
@@ -54,32 +54,134 @@ inline Node mkValidPhase(TNode a, TNode pi)
...
 void SineSolver::doPhaseShift(TNode a, TNode new_a, TNode y)
 {
   NodeManager* nm = NodeManager::currentNM();
   SkolemManager* sm = nm->getSkolemManager();
   Assert(a.getKind() == Kind::SINE);
   Trace("nl-ext-tf") << "Basis sine : " << new_a << " for " << a << std::endl;
-  Assert(!d_data->d_pi.isNull());
   Node shift = sm->mkDummySkolem("s", nm->integerType(), "number of shifts");
   // TODO (cvc4-projects #47) : do not introduce shift here, instead needs model-based
   // refinement for constant shifts (cvc4-projects #1284)
```
Comment (nafur) [inline src/theory/arith/nl/transcendental/sine_solver.cpp:172]: Update the comment

File: src/theory/arith/nl/transcendental/sine_solver.cpp [hunk 2]
```diff
@@ -54,32 +54,134 @@ inline Node mkValidPhase(TNode a, TNode pi)
...
+  {
+    return;
+  }
+  std::vector<Node> mpvs;
+  for (const Node& m : d_mpoints)
+  {
+    mpvs.push_back(d_data->d_model.computeAbstractModelValue(m));
+  }
+  std::map<Node, Node> valForSym;
+  std::map<Node, Node>::iterator itv;
```
Comment (nafur) [inline src/theory/arith/nl/transcendental/sine_solver.cpp:96]: only declare within the loop.

File: src/theory/arith/nl/transcendental/sine_solver.cpp [hunk 3]
```diff
@@ -54,32 +54,134 @@ inline Node mkValidPhase(TNode a, TNode pi)
...
+        Node lem =
+            nm->mkNode(kind::IMPLIES,
+                       tf[0].eqNode(nm->mkNode(kind::NEG, itv->second[0])),
+                       tf.eqNode(nm->mkNode(kind::NEG, itv->second)));
+        d_data->d_im.addPendingLemma(
+            lem, InferenceId::ARITH_NL_T_SINE_SYMM, nullptr);
+      }
+      // we do not consider it reduced currently, since we require setting
+      // approximate bounds for it, alternatively we could carry the negation
+      // of the approximation in the transcendental solver
```
Comment (nafur) [inline src/theory/arith/nl/transcendental/sine_solver.cpp:119]: I would prefer to `continue;` right here, simplifying the structure of the loop body.

Comment (ajreynol) [reply src/theory/arith/nl/transcendental/sine_solver.cpp:119]: We require falling through so that it is added to `nreduced` at the bottom of the loop.

**Review by nafur — APPROVED** [review]
LGTM.

**Pull request merged**